File: ast/types/type_pointer.go
package types

import (
	"fmt"

	"github.com/panda-io/micro-panda/ast/core"
)

type TypePointer struct {
	TypeBase
	ElementType core.Type
}

func (p *TypePointer) Equal(t core.Type) bool {
	if pointer, ok := t.(*TypePointer); ok {
		return p.ElementType.Equal(pointer.ElementType)
	} else if array, ok := t.(*TypeArray); ok {
		if len(array.Dimension) == 1 {
			return p.ElementType.Equal(array.ElementType)
		}
	}
	return false
}

func (p *TypePointer) String() string {
	return fmt.Sprintf("pointer<%s>", p.ElementType.String())
}

File: app/src/main/java/com/github/panda_io/micro_panda/ast/type/TypePointer.java
package com.github.panda_io.micro_panda.ast.type;

public class TypePointer extends Type {
    public Type elementType;
    public boolean isRaw;

    public TypePointer() {
    }

    public boolean equal(Type type) {
        if (type == null)
            return false;
        if (type instanceof TypePointer) {
            if (((TypePointer) type).isRaw || this.isRaw) {
                return true;
            }
            return this.elementType.equal(((TypePointer) type).elementType);
        } else if (type instanceof TypeArray) {
            if (((TypeArray) type).dimensions.size() == 1) {
                return this.elementType.equal(((TypeArray) type).elementType);
            }
        }
        return false;
    }

    public String string() {
        return String.format("pointer<%s>", this.elementType.string());
    }
}
